cmake/SYCL.cmake

```diff
@@ -91,7 +95,40 @@ else()
         list(APPEND EXTRA_SHARED_LIBS OpenCL::OpenCL)
     endif()
 endif()
-
+if(DNNL_SYCL_HIP)
```

Comment:
The way HIP backend in integrated in the oneDNN build system has a significant drawback due to using CMake module files for HIP, rocBLAS and MIOpen from the ROCm package. The problem is that those module files seem to rely on the fact that a Clang compiler from ROCm package will be used, which is not the case for us. For example, some libraries are missing in DPC++ compiler, which causes a link time error. Also, the modules add many compile flags, which are not required and may cause other errors as well. So I made some experiments with custom CMake module for all the needed dependencies.

The CMake module files that should be placed to `cmake` directory located in oneDNN directory tree:

<details>
  <summary>FindrocBLAS.cmake</summary>

```cmake
find_package(HIP REQUIRED)
find_package(Threads REQUIRED)

# Prioritize ROCBLASROOT
list(APPEND rocblas_root_hints
            ${ROCBLASROOT}
            $ENV{ROCBLASROOT}
            "/opt/rocm/rocblas")

find_path(
    rocBLAS_INCLUDE_DIR "rocblas.h"
    HINTS ${rocblas_root_hints}
    PATH_SUFFIXES include
)

find_library(
    rocBLAS_LIBRARY rocblas
    HINTS ${rocblas_root_hints}
    PATH_SUFFIXES lib
)

if(EXISTS "${rocBLAS_INCLUDE_DIR}/internal/rocblas-version.h")
    file(READ "${rocBLAS_INCLUDE_DIR}/internal/rocblas-version.h" rocBLAS_VERSION_CONTENT)

    string(REGEX MATCH "define ROCBLAS_VERSION_MAJOR +([0-9]+)" _ "${rocBLAS_VERSION_CONTENT}")
    set(rocBLAS_MAJOR_VERSION ${CMAKE_MATCH_1} CACHE INTERNAL "")

    string(REGEX MATCH "define ROCBLAS_VERSION_MINOR +([0-9]+)" _ "${rocBLAS_VERSION_CONTENT}")
    set(rocBLAS_MINOR_VERSION ${CMAKE_MATCH_1} CACHE INTERNAL "")

    string(REGEX MATCH "define ROCBLAS_VERSION_PATCH +([0-9]+)" _ "${rocBLAS_VERSION_CONTENT}")
    set(rocBLAS_PATCH_VERSION ${CMAKE_MATCH_1} CACHE INTERNAL "")

    set(rocBLAS_VERSION
        "${rocBLAS_MAJOR_VERSION}.${rocBLAS_MINOR_VERSION}.${rocBLAS_PATCH_VERSION}"
    )

    unset(rocBLAS_VERSION_CONTENT)
else()
    message(WARNING "rocBLAS version couldn't be identified.")
endif()

include(FindPackageHandleStandardArgs)

find_package_handle_standard_args(rocBLAS
    FOUND_VAR rocBLAS_FOUND
    REQUIRED_VARS
        rocBLAS_LIBRARY
        rocBLAS_INCLUDE_DIR
    VERSION_VAR rocBLAS_VERSION
)

if(rocBLAS_FOUND AND NOT TARGET rocBLAS::rocBLAS)
    add_library(rocBLAS::rocBLAS SHARED IMPORTED)
    set_target_properties(rocBLAS::rocBLAS PROPERTIES
        IMPORTED_LOCATION "${rocBLAS_LIBRARY}"
        INTERFACE_INCLUDE_DIRECTORIES "${rocBLAS_INCLUDE_DIR}"
        INTERFACE_LINK_LIBRARIES "HIP::HIP;Threads::Threads"
    )
endif()
```

</details>

<details>
  <summary>FindMIOpen.cmake</summary>

```cmake
find_package(HIP REQUIRED)
find_package(Threads REQUIRED)
find_package(rocBLAS REQUIRED)

# Rely on the standard CMake config for amd_comgr as it doesn't add redundant
# dependencies.
find_package(amd_comgr REQUIRED CONFIG
    HINTS ${COMGRROOT} $ENV{COMGRROOT} /opt/rocm/lib/cmake
)
# XXX: find a way to propagate it via target properties.
list(APPEND EXTRA_SHARED_LIBS amd_comgr)

# Prioritize MIOPENROOT
list(APPEND miopen_root_hints
            ${MIOPENROOT}
            $ENV{MIOPENROOT}
            "/opt/rocm/miopen"
)

find_path(
    MIOpen_INCLUDE_DIR "miopen/miopen.h"
    HINTS ${miopen_root_hints}
    PATH_SUFFIXES include
)

find_library(
    MIOpen_LIBRARY MIOpen
    HINTS ${miopen_root_hints}
    PATH_SUFFIXES lib
)

if(EXISTS "${MIOpen_INCLUDE_DIR}/miopen/version.h")
    file(READ "${MIOpen_INCLUDE_DIR}/miopen/version.h" MIOpen_VERSION_CONTENT)

    string(REGEX MATCH "define MIOPEN_VERSION_MAJOR +([0-9]+)" _ "${MIOpen_VERSION_CONTENT}")
    set(MIOpen_MAJOR_VERSION ${CMAKE_MATCH_1} CACHE INTERNAL "")

    string(REGEX MATCH "define MIOPEN_VERSION_MINOR +([0-9]+)" _ "${MIOpen_VERSION_CONTENT}")
    set(MIOpen_MINOR_VERSION ${CMAKE_MATCH_1} CACHE INTERNAL "")

    string(REGEX MATCH "define MIOPEN_VERSION_PATCH +([0-9]+)" _ "${MIOpen_VERSION_CONTENT}")
    set(MIOpen_PATCH_VERSION ${CMAKE_MATCH_1} CACHE INTERNAL "")

    set(MIOpen_VERSION
        "${MIOpen_MAJOR_VERSION}.${MIOpen_MINOR_VERSION}.${MIOpen_PATCH_VERSION}"
    )

    unset(MIOpen_VERSION_CONTENT)
else()
    message(WARNING "MIOpen version couldn't be identified.")
endif()

include(FindPackageHandleStandardArgs)

find_package_handle_standard_args(MIOpen
    FOUND_VAR MIOpen_FOUND
    REQUIRED_VARS
        MIOpen_LIBRARY
        MIOpen_INCLUDE_DIR
    VERSION_VAR MIOpen_VERSION
)

if(MIOpen_FOUND AND NOT TARGET MIOpen::MIOpen)
    add_library(MIOpen::MIOpen SHARED IMPORTED)
    set_target_properties(MIOpen::MIOpen PROPERTIES
        IMPORTED_LOCATION "${MIOpen_LIBRARY}"
        INTERFACE_INCLUDE_DIRECTORIES "${MIOpen_INCLUDE_DIR}"
        INTERFACE_LINK_LIBRARIES "HIP::HIP;rocBLAS::rocBLAS;amd_comgr;Threads::Threads"
    )
endif()
```

</details>

<details>
  <summary>FindHIP.cmake</summary>

```cmake
find_package(Threads REQUIRED)

# Prioritize HIPROOT
list(APPEND hip_root_hints
            ${HIPROOT}
            $ENV{HIPROOT}
            "/opt/rocm/hip")

find_path(
    HIP_INCLUDE_DIR "hip/hip_runtime_api.h"
    HINTS ${hip_root_hints}
    PATH_SUFFIXES include
)

find_library(
    HIP_LIBRARY amdhip64
    HINTS ${hip_root_hints}
    PATH_SUFFIXES lib
)

if(EXISTS "${HIP_INCLUDE_DIR}/hip/hip_version.h")
    file(READ "${HIP_INCLUDE_DIR}/hip/hip_version.h" HIP_VERSION_CONTENT)

    string(REGEX MATCH "define HIP_VERSION_MAJOR +([0-9]+)" _ "${HIP_VERSION_CONTENT}")
    set(HIP_MAJOR_VERSION ${CMAKE_MATCH_1} CACHE INTERNAL "")

    string(REGEX MATCH "define HIP_VERSION_MINOR +([0-9]+)" _ "${HIP_VERSION_CONTENT}")
    set(HIP_MINOR_VERSION ${CMAKE_MATCH_1} CACHE INTERNAL "")

    string(REGEX MATCH "define HIP_VERSION_PATCH +([0-9]+)" _ "${HIP_VERSION_CONTENT}")
    set(HIP_PATCH_VERSION ${CMAKE_MATCH_1} CACHE INTERNAL "")

    set(HIP_VERSION
        "${HIP_MAJOR_VERSION}.${HIP_MINOR_VERSION}.${HIP_PATCH_VERSION}"
    )

    unset(HIP_VERSION_CONTENT)
else()
    message(WARNING "HIP version couldn't be identified.")
endif()

include(FindPackageHandleStandardArgs)

find_package_handle_standard_args(HIP
    FOUND_VAR HIP_FOUND
    REQUIRED_VARS
        HIP_LIBRARY
        HIP_INCLUDE_DIR
    VERSION_VAR HIP_VERSION
)

if(HIP_FOUND AND NOT TARGET HIP::HIP)
    add_library(HIP::HIP SHARED IMPORTED)
    set_target_properties(HIP::HIP PROPERTIES
        IMPORTED_LOCATION "${HIP_LIBRARY}"
        INTERFACE_INCLUDE_DIRECTORIES "${HIP_INCLUDE_DIR}"
        INTERFACE_LINK_LIBRARIES "Threads::Threads"
        INTERFACE_COMPILE_DEFINITIONS "__HIP_PLATFORM_AMD__=1"
    )
endif()

```

</details>

Once the files are in the `cmake` directory we can simply add a couple lines in `SYCL.cmake`:
```cmake
find_package(rocBLAS REQUIRED)
find_package(MIOpen REQUIRED)
````
And change names of the targets in `src/gpu/amd/CMakeLists.txt` accordingly:

- `roc::rocblas` -> `rocBLAS::rocBLAS`
- `MIOpen` -> `MIOpen::MIOpen`

I tested it on my side and it worked. I find this approach better for us as we have full control over dependencies. Please let me know if you see any issue with it.